python/ray/data/_internal/execution/operators/actor_pool_submitter.py

```diff
@@ -0,0 +1,179 @@
+        ray_remote_args = ray_remote_args.copy()
+        if "num_cpus" not in ray_remote_args:
+            ray_remote_args["num_cpus"] = 1
+        if "scheduling_strategy" not in ray_remote_args:
+            ctx = DatasetContext.get_current()
+            if ctx.scheduling_strategy == DEFAULT_SCHEDULING_STRATEGY:
+                ray_remote_args["scheduling_strategy"] = "SPREAD"
+            else:
+                ray_remote_args["scheduling_strategy"] = ctx.scheduling_strategy
+        return ray_remote_args
+
+
+class MapWorker:
+    """An actor worker for MapOperator."""
+
+    def ready(self):
+        return "ok"
+
+    def submit(
+        self, fn: Callable[[Iterator[Block]], Iterator[Block]], *blocks: Block
+    ) -> Tuple[Block, BlockMetadata]:
+        # Coalesce all fn output blocks.
+        # TODO(Clark): Remove this coalescing once dynamic block splitting is supported
+        # for actors.
+        # yield from _map_task(fn, *blocks)
+        stats = BlockExecStats.builder()
+        builder = DelegatingBlockBuilder()
+        for block in fn(iter(blocks)):
+            builder.add_block(block)
+        block = builder.build()
+        block_metadata = BlockAccessor.for_block(block).get_metadata([], None)
+        block_metadata.exec_stats = stats.build()
+        return block, block_metadata
+
+
+class ActorPool:
+    """A pool of actors for map task execution.
+
+    This class is in charge of tracking the number of in-flight tasks per actor,
+    providing the least heavily loaded actor to the task submitter, and killing idle
+    actors when the task submitter is done submitting work to the pool.
+    """
+
+    def __init__(self):
+        # Number of tasks in flight per actor.
+        self._tasks_in_flight: Dict[ray.actor.ActorHandle, int] = {}
+        # Whether actors that become idle should be eagerly killed. This is False until
+        # the first call to kill_idle_actors().
+        self._should_kill_idle_actors = False
+
+    def add_actor(self, actor: ray.actor.ActorHandle):
+        """Adds an actor to the pool."""
+        self._tasks_in_flight[actor] = 0
+
+    def pick_actor(self) -> ray.actor.ActorHandle:
+        """Provides the least heavily loaded actor in the pool for task submission."""
+        actor = min(
+            self._tasks_in_flight.keys(), key=lambda actor: self._tasks_in_flight[actor]
+        )
+        self._tasks_in_flight[actor] += 1
+        return actor
+
+    def return_actor(self, actor: ray.actor.ActorHandle):
+        """Returns the provided actor to the pool."""
+        self._tasks_in_flight[actor] -= 1
+        if self._should_kill_idle_actors and self._tasks_in_flight[actor] == 0:
+            self._kill_actor(actor)
+
+    def has_actor(self, actor: ray.actor.ActorHandle) -> bool:
+        """Whether the provided actor is in this pool."""
+        return actor in self._tasks_in_flight
+
+    def get_tasks_in_flight(self, actor: ray.actor.ActorHandle) -> int:
```

Comment:
nit: get_num_tasks_in_flight() or just num_tasks_in_flight(), since this is not returning tasks

Reply:
Ended up removing this method after rewriting the tests to not focus on internal state.